lib/WUI/espif.c

```diff
@@ -0,0 +1,521 @@
+                    state = Intron;
+                }
+            }
+            break;
+
+        case PacketData: {
+            // Copy input to current pbuf (until end of input or current pbuf)
+            const uint32_t to_read = min(rx_buff_cur->len - rx_read, end - c);
+            memcpy(rx_buff_cur->payload + rx_read, c, to_read);
+            c += to_read;
+            rx_read += to_read;
+
+            // Switch to next pbuf
+            if (rx_read == rx_buff_cur->len) {
+                rx_buff_cur = rx_buff_cur->next;
+                rx_read = 0;
+            }
+
+            // Filled all pbufs in a packet (current set to next = NULL)
+            if (!rx_buff_cur) {
+                log_debug(ESPIF, "Read packet size: %d", rx_len);
+                if (netif->input(rx_buff, netif) != ERR_OK) {
+                    log_error(ESPIF, "ethernetif_input: IP input error");
+                    pbuf_free(rx_buff);
+                    state = Intron;
+                    break;
+                }
+                log_debug(ESPIF, "Input packet processed ok");
+                state = Intron;
+            }
+        } break;
+        default:
+            log_error(ESPIF, "Unknow input parser state: %d", state);
+            state = Intron;
+        }
+    }
+    log_debug(ESPIF, "Processed %d from UART", size);
+}
+
+static err_t low_level_output(struct netif *netif, struct pbuf *p) {
+    const uint32_t len = p->tot_len;
+    log_debug(ESPIF, "Low level output packet size: %d", len);
+
+    espif_transmit_data(INTRON, sizeof(INTRON));
+    uint8_t msg_type = MSG_PACKET;
+    espif_transmit_data(&msg_type, 1);
+    espif_transmit_data(&len, sizeof(len));
+    while (p != NULL) {
+        int written = espif_transmit_data(p->payload, p->len);
+        if (written != p->len) {
+            log_error(ESPIF, "Low level output packet failed %d != %d", p->len, written);
+            return ERR_IF;
+        }
+        p = p->next;
+    }
+
+    return ERR_OK;
+}
+
+err_t espif_init(struct netif *netif) {
+    log_info(ESPIF, "LwIP init");
+    low_level_init(netif);
+
+    // Initialize lwip netif
+    netif->name[0] = 'w';
+    netif->name[1] = 'l';
+    netif->output = etharp_output;
+#if LWIP_IPV6
+    netif->output_ip6 = ethip6_output;
+#endif
+    netif->linkoutput = low_level_output;
+
+    // LL init
+    netif->hwaddr_len = 0;
+    // TODO: This assumes LwIP can live with hwaddr not being set until ESP reports it
+    netif->mtu = 1500;
+    netif->flags = NETIF_FLAG_BROADCAST | NETIF_FLAG_ETHARP;
+
+    // Reset device to receive MAC address
+    log_debug(ESPIF, "Resetting ESP to receive MAC address");
+    hard_reset_device();
+
+    // Wait for esp ready (receiveing device info)
+    init_task_handle = xTaskGetCurrentTaskHandle();
+    if (!xTaskNotifyWait(0, 0, NULL, pdMS_TO_TICKS(3000))) {
+        log_error(ESPIF, "Timeout waiting for ESP bootup");
+        low_level_deinit();
+        return ERR_IF;
```

Comment:
It can happen it ESP does not send the devinfo message for some reason. The reasons are many from missing ESP to borked firmware. When this happens netif init fails and it stays down forever.